### PR TITLE
Validate SAML2 bearer assertions with SAML SP's IdP Entity ID Alias as issuer

### PR DESCRIPTION
**Pull request by DinikaSen**
### Proposed changes in this pull request
In SAML 2.0 Bearer Assertion Profile, the SAML 2.0 assertion received by the OAuth server is validated using the "issuer" of the SAML assertion. With a new feature introduced to SAML SSO inbound configuration to allow overriding IdP Entity ID per SAML SP, the validation of SAML assertion issuer has to be done against the SAML SP's IdP Entity ID Alias. This PR adds this new validation part to the "SAML2BearerGrantHadler" class.

### When should this PR be merged
This PR should be merged after merging the following 2 PRs -
wso2/carbon-identity-framework/pull/1999
wso2-extensions/identity-inbound-auth-saml/pull/201

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/saml/SAML2BearerGrantHandler.java [hunk 1]
```diff
@@ -798,7 +803,7 @@ private IdentityProvider getIdentityProvider(Assertion assertion, String tenantD
     }
 
     private IdentityProvider getIdentityProviderFromManager(Assertion assertion, String tenantDomain)
-            throws IdentityProviderManagementException {
+            throws IdentityProviderManagementException, IdentityOAuth2Exception {
```
Comment (ayshsandu) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/saml/SAML2BearerGrantHandler.java]: Why do we need to throw a new type of exception?

**Comment by malithie** [comment]
@ayshsandu , this seems to be related to a specific feature.
Can we pls get a feature branch created and merge this PR there

**Comment by dewniMW** [comment]
I've locally merged the PR and tested SAML2 Bearer Assertion Profile for OAuth 2.0 flow. There are no merge conflicts.

**Review by ayshsandu — APPROVED** [review]

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/saml/SAML2BearerGrantHandler.java [hunk 2]
```diff
@@ -835,6 +850,88 @@ private void checkNullIdentityProvider(Assertion assertion, String tenantDomain,
...
+        }
+        if (log.isDebugEnabled()) {
+            log.debug("No SAML Service Provider configuration with IdP Entity ID Alias " +
+                    "similar to token issuer found.");
+        }
+        return false;
+    }
+
+    private SAMLSSOServiceProviderDO getSAMLSSOServiceProvider(String issuerName, String tenantDomain)
+            throws IdentityOAuth2Exception {
```
Comment (Kanapriya) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/saml/SAML2BearerGrantHandler.java:899]: Add some method-level comments

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/saml/SAML2BearerGrantHandler.java [hunk 3]
```diff
@@ -732,7 +737,7 @@ private String getTokenEPAliasFromResidentIdp(Assertion assertion, IdentityProvi
                                                   String tenantDomain) throws IdentityOAuth2Exception {
         String tokenEndpointAlias = null;
         FederatedAuthenticatorConfig[] fedAuthnConfigs = identityProvider.getFederatedAuthenticatorConfigs();
-        validateIdpEntityId(assertion, tenantDomain,  getIdpEntityId(fedAuthnConfigs));
+        //validateIdpEntityId(assertion, tenantDomain,  getIdpEntityId(fedAuthnConfigs));
```
Comment (sachiniWettasinghe) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/saml/SAML2BearerGrantHandler.java:740]: Why have we commented this line? Can't we remove it?

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/saml/SAML2BearerGrantHandlerTest.java [hunk 4]
```diff
@@ -228,8 +232,8 @@ public void testValidateGrant(String userType, String idpName) throws Exception
                 { validOnOrAfter, "LOCAL", true, true, "invalidAudience", TestConstants.LOACALHOST_DOMAIN, null, "Audience Restriction validation failed"},
                 { validOnOrAfter, "LOCAL", true, true, "", TestConstants.LOACALHOST_DOMAIN, null, "Token Endpoint alias has not been configured"},
                 { validOnOrAfter, "FED", true, true, "invalidAudience", TestConstants.LOACALHOST_DOMAIN, null, "Audience Restriction validation failed"},
-                { validOnOrAfter, "LOCAL", false, true, TestConstants.OAUTH2_TOKEN_EP, "invalidEntityId", null, "Issuer verification failed"},
-                { validOnOrAfter, "LOCAL", false, true, TestConstants.OAUTH2_TOKEN_EP, null, null, "Issuer verification failed"},
+//                { validOnOrAfter, "LOCAL", false, true, TestConstants.OAUTH2_TOKEN_EP, "invalidEntityId", null, "Issuer verification failed"},
+//                { validOnOrAfter, "LOCAL", false, true, TestConstants.OAUTH2_TOKEN_EP, null, null, "Issuer verification failed"},
```
Comment (sachiniWettasinghe) [inline components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/saml/SAML2BearerGrantHandlerTest.java:236]: Why have commented these?

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/saml/SAML2BearerGrantHandlerTest.java [hunk 5]
```diff
@@ -260,6 +265,7 @@ public void testValidateGrantException(Object dateTimeObj, String idpName, boole
         when(userRealm.getUserStoreManager()).thenReturn(userStoreManager);
         when(userStoreManager.isExistingUser(anyString())).thenReturn(isUserExist);
 
+
```
Comment (sachiniWettasinghe) [inline components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/saml/SAML2BearerGrantHandlerTest.java:268]: Remove redundant line

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/saml/SAML2BearerGrantHandlerTest.java [hunk 6]
```diff
@@ -485,4 +491,24 @@ public void handleSAML2Token(OAuthTokenReqMessageContext tokReqMsgCtx) throws Id
         when(oAuthServerConfiguration.getSAML2TokenCallbackHandler()).thenReturn(callbackHandler);
     }
 
+    private void prepareForGetSAMLSSOServiceProvider() throws Exception {
```
Comment (sachiniWettasinghe) [inline components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/saml/SAML2BearerGrantHandlerTest.java:494]: Add new line after method signature

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/saml/SAML2BearerGrantHandler.java (hunk 2 above)
Comment (DinikaSen) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/saml/SAML2BearerGrantHandler.java:899]: Addressed in https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1438

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/saml/SAML2BearerGrantHandler.java (hunk 3 above)
Comment (DinikaSen) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/saml/SAML2BearerGrantHandler.java:740]: Addressed in https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1438

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/saml/SAML2BearerGrantHandlerTest.java (hunk 4 above)
Comment (DinikaSen) [reply components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/saml/SAML2BearerGrantHandlerTest.java:236]: This section was removed by a later PR

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/saml/SAML2BearerGrantHandlerTest.java (hunk 5 above)
Comment (DinikaSen) [reply components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/saml/SAML2BearerGrantHandlerTest.java:268]: Addressed in https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1438

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/saml/SAML2BearerGrantHandlerTest.java (hunk 6 above)
Comment (DinikaSen) [reply components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/saml/SAML2BearerGrantHandlerTest.java:494]: Addressed in https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1438

**Pull request merged**